### PR TITLE
Remove write_predictions from LightningModule

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

`write_predictions` was marked as deprecated in v1.3 and is being removed in v1.5. Users can write outputs directly from their lightning module to their preferred destination, or use the prediction writer callbacks now offered. 

This contributes to https://github.com/PyTorchLightning/pytorch-lightning/issues/7315 as better hygiene and separation across components.

### Does your PR introduce any breaking changes? If yes, please list them.
Remove `write_predictions` and `write_predictions_dict` from LightningModule.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

File: CHANGELOG.md [hunk 1]
```diff
@@ -118,6 +118,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Removed the deprecated `sync_step` argument from `WandbLogger` ([#8763](https://github.com/PyTorchLightning/pytorch-lightning/pull/8763))
 
 
+- Removed `LightningModule.write_predictions` and `LightningModule.write_predictions_dict` ([#8850](https://github.com/PyTorchLightning/pytorch-lightning/pull/8850/))
```
Comment (awaelchli) [inline CHANGELOG.md:121]:
```suggestion
- Removed `LightningModule.write_predictions` and `LightningModule.write_predictions_dict` in favor of `Trainer.predict` and `BasePredictionWriter` ([#8850](https://github.com/PyTorchLightning/pytorch-lightning/pull/8850/))
```

Comment (awaelchli) [reply CHANGELOG.md:121]: what do you think, should we mention these alternatives?

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 2]
```diff
@@ -198,9 +194,6 @@ def on_evaluation_batch_end(
 
         self.trainer.logger_connector.on_batch_end()
 
-        # store predicitons if do_write_predictions and track eval loss history
-        self.store_predictions(output, batch_idx, dataloader_idx)
-
     def store_predictions(self, output: Optional[STEP_OUTPUT], batch_idx: int, dataloader_idx: int) -> None:
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:197]: this is dead code now, can also be removed. but there is a call there to the dev debugger which we need to pull up and keep.

Comment (awaelchli) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:197]: couldn't push to your branch due to permission error, but removing lines 197 - 204 will fix the tests

Comment (ananthsub) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py]: thanks @awaelchli !

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 3]
```diff
@@ -19,9 +19,7 @@
 from torch import Tensor
 
 from pytorch_lightning.loops.base import Loop
-from pytorch_lightning.trainer.connectors.logger_connector.result import ResultCollection
 from pytorch_lightning.trainer.progress import Progress
-from pytorch_lightning.trainer.supporters import PredictionCollection
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:24]: I believe `PredictionCollection` is now unused and can be removed

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 4]
```diff
@@ -112,8 +108,8 @@ def advance(
 
         self.batch_progress.increment_processed()
 
-        # hook + store predictions
-        self.on_evaluation_batch_end(output, batch, batch_idx, dataloader_idx)
+        # track loss history
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:111]: You accidentally removed this

```suggestion
        self.on_evaluation_batch_end(output, batch, batch_idx, dataloader_idx)

        # track loss history
```

**Pull request merged**